hubspot/plugin.py

```diff
@@ -135,57 +478,61 @@ def create_contacts(contacts: List[HubSpotContact]) -> List[str]:
 
 
 @dataclass
-class HubSpotContactUpdate:
-    """The data required to update a HubSpot contact.
+class ContactDefaultPropertyUpdate:
+    """The data required to update a contact's default properties."""
 
-    Attributes:
-        id: The HubSpot ID of the contact to update.
-        updated_properties: A dictionary of property names and values to update for the
-            contact. Only properties that are to be updated need to be included.
+    contact_id: str
+    updated_properties: List[Tuple[HubSpotContactDefaultPropertyName, str]]
 
-    """
 
-    id: str
-    updated_properties: Dict[str, str]
+@dataclass
+class ContactCustomPropertyUpdate:
+    """The data required to update a contact's custom properties."""
+
+    contact_id: str
+    updated_properties: List[Tuple[HubSpotContactCustomPropertyName, str]]
 
 
-def update_contacts(contacts: List[HubSpotContactUpdate]) -> List[str]:
+def update_contacts(
+    default_properties: List[ContactDefaultPropertyUpdate],
+    custom_properties: List[ContactCustomPropertyUpdate],
+) -> List[str]:
     """
     Update multiple contacts in HubSpot.
 
-    Args:
-        contacts: A list of HubSpotContactUpdate objects to be updated.
-
     Returns:
-        A list of strings, where each string is the HubSpot ID of an updated contact.
+        Contact IDs that have been updated.
     """
     url = "https://api.hubapi.com/crm/v3/objects/contacts/batch/update"
 
-    # Prepare the payload from the contacts list
-    contacts_payload = []
-    for contact in contacts:
-        contact_data = {
-            "id": contact.id,  # Contact ID is required for updating
-            "properties": contact.updated_properties,
-        }
-        contacts_payload.append(contact_data)
+    merged_properties = defaultdict(dict)
 
-    payload = {"inputs": contacts_payload}
+    for property_update in default_properties + custom_properties:
+        merged_properties[property_update.contact_id].update(
+            {prop.name: value for prop, value in property_update.updated_properties}
+        )
+
+    payload = [
+        {"id": contact_id, "properties": properties}
+        for contact_id, properties in merged_properties.items()
+    ]
 
     with httpx.Client(
         transport=AugmentedTransport(actions_v0.authenticated_request_hubspot),
     ) as client:
-        response = client.post(url, json=payload)
+        response = client.post(url, json={"inputs": payload})
         response.raise_for_status()
         data = response.json()
-
-    # Extract and return the IDs of the updated contacts
-    return [result["id"] for result in data["results"]]
+        return [result["id"] for result in data["results"]]
 
 
 def search_contacts(
     search_criteria: Dict[str, str],
-    additional_properties: Optional[Sequence[str]] = None,
+    additional_properties: Optional[
+        Sequence[
+            Union[HubSpotContactDefaultPropertyName, HubSpotContactCustomPropertyName]
+        ]
+    ] = None,
```

Comment:
Unions are good here? (if so, this is ok)

generally, Unions as args i think are fine.

Unions for return types are bad

Reply:
Yes. This seems to work well.  For example when using a prompt:
```
I have a hubspot contact that i can look up by email. Please update the contact's hubspot property "lutrauserid" for me.
```
code_gen recognizes that as not part of the default property list and generates code that looks like this:

```
def update_hubspot_contact_lutrauserid(email: str, lutrauserid: str) -> OutputMarkdown:
    # Search for the contact by email 
    try:
        contacts: Sequence[HubSpotContact] = hubspot_search_contacts({"email": email}, [HubSpotContactCustomPropertyName(name="lutrauserid")])
    except Exception as e:
        raise RuntimeError(f"Error searching for contact: {str(e)}") from e

    if len(contacts) == 0:
        # No contact found with the provided email
        result_message: str = f"No HubSpot contact found with email: {email}"
    else:
        # Contact found, prepare the update payload
        contact: HubSpotContact = contacts[0]
        updated_properties: List[Tuple[HubSpotContactCustomPropertyName, str]] = [(HubSpotContactCustomPropertyName(name="lutrauserid"), lutrauserid)]
        update: ContactCustomPropertyUpdate = ContactCustomPropertyUpdate(contact_id=contact.id, updated_properties=updated_properties)

        # Update the contact
        try:
            updated_contact_ids: Sequence[str] = hubspot_update_contacts([], [update])
            result_message = f"Successfully updated HubSpot contact with email {email}. Set lutrauserid to {lutrauserid}."
        except Exception as e:
            raise RuntimeError(f"Error updating contact: {str(e)}") from e

    # Send the result message
    try:
        send_email_to_current_user("HubSpot Contact Update Result", MarkdownDocument(title="HubSpot Contact Update Result", content=result_message))
    except Exception as e:
        raise RuntimeError(f"Error sending email: {str(e)}") from e

    return OutputMarkdown(markdown=result_message)
```